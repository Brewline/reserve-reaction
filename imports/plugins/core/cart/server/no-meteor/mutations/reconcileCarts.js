import Logger from "@reactioncommerce/logger";
import ReactionError from "@reactioncommerce/reaction-error";
import hashLoginToken from "/imports/node-app/core/util/hashLoginToken";
import convertAnonymousCartToNewAccountCart from "./convertAnonymousCartToNewAccountCart";
import reconcileCartsKeepAccountCart from "./reconcileCartsKeepAccountCart";
import reconcileCartsKeepAnonymousCart from "./reconcileCartsKeepAnonymousCart";
import reconcileCartsMerge from "./reconcileCartsMerge";

/**
 * @method reconcileCarts
 * @summary Call this with account credentials, passing in an anonymous cart, and the
 *   anonymous cart will be merged into the account cart. The "mode" argument allows
 *   you to specify whether the items should be merged, or if items should be kept from
 *   just one of the carts. If this mutation does not throw an error, the anonymous cart
 *   will be destroyed by the time this function returns.
 * @param {Object} context - an object containing the per-request state
 * @param {Object} input - mutation input
 * @param {String} input.anonymousCartId - The anonymous cart ID
 * @param {String} input.anonymousCartToken - The anonymous cart token
 * @param {String} [input.mode] - The reconciliation mode, "merge", "keepAccountCart", or "keepAnonymousCart". Default "merge"
 * @return {Promise<Object>} Object in which `cart` property is set to the updated account cart
 */
export default async function reconcileCarts(context, input) {
  const { accountId, collections, user, userId = null } = context;
  const { Cart } = collections;
  const { anonymousCartId, anonymousCartToken, mode = "merge" } = input;

  if (!accountId) throw new ReactionError("access-denied", "Access Denied");
  if (!anonymousCartId) throw new ReactionError("invalid-param", "anonymousCartId is required");
  if (!anonymousCartToken) throw new ReactionError("invalid-param", "anonymousCartToken is required");

  const accountCartSelector = { accountId };
  const anonymousCartSelector = { _id: anonymousCartId, anonymousAccessToken: hashLoginToken(anonymousCartToken) };

  const carts = await Cart.find({
    $or: [accountCartSelector, anonymousCartSelector]
  }).toArray();

  const anonymousCart = carts.find((cart) => cart._id === anonymousCartId);
  if (!anonymousCart) throw new ReactionError("not-found", "Anonymous cart not found");

  const { shopId } = anonymousCart;

  // In the Meteor app, there are accounts for anonymous users. This check can be removed someday.
  // Don't use `userHasPermission` for this check because that always returns true if there
  // is "owner" role. We want to know explicitly whether they have the "anonymous" role.
  const roles = (user.roles && user.roles[shopId]) || [];
  if (roles.indexOf("anonymous") !== -1) {
    Logger.warn("reconcileCarts called by an anonymous user. Check client code.");
    throw new ReactionError("access-denied", "Access Denied");
  }

  const accountCart = carts.find((cart) => cart.accountId === accountId && cart.shopId === shopId);

  if (accountCart) {
    // We have both carts, so reconcile them according to "mode"
    switch (mode) {
      case "keepAccountCart":
        return {
          cart: await reconcileCartsKeepAccountCart({ accountCart, anonymousCartSelector, Cart, userId })
        };

      case "keepAnonymousCart":
        return {
          cart: await reconcileCartsKeepAnonymousCart({ accountCart, accountCartSelector, anonymousCart, anonymousCartSelector, Cart, userId })
        };

      case "merge":
        return {
          cart: await reconcileCartsMerge({ accountCart, accountCartSelector, anonymousCart, anonymousCartSelector, context, userId })
        };

      default:
        throw new ReactionError("invalid-param", "mode must be keepAccountCart, keepAnonymousCart, or merge");
    }
  }

  // We have only an anonymous cart, so convert it to an account cart
  return {
    cart: await convertAnonymousCartToNewAccountCart({
      accountId,
      anonymousCart,
      anonymousCartSelector,
      Cart,
      shopId,
      userId
    })
  };
}
