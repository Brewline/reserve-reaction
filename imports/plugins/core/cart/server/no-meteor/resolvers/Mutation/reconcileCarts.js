import { decodeCartOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/cart";

/**
 * @name "Mutation.reconcileCarts"
 * @method
 * @memberof Cart/GraphQL
 * @summary resolver for the reconcileCarts GraphQL mutation
 * @param {Object} parentResult - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} args.input.anonymousCartId - The anonymous cart ID
 * @param {String} args.input.anonymousCartToken - The anonymous access token that was returned from `createCart`
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {String} [args.input.mode] - The reconciliation mode, "merge", "keepAccountCart", or "keepAnonymousCart". Default "merge"
 * @param {Object} context - an object containing the per-request state
 * @return {Promise<Object>} ReconcileCartsPayload
 */
export default async function reconcileCarts(parentResult, { input }, context) {
  const {
    anonymousCartId: opaqueAnonymousCartId,
    anonymousCartToken,
    clientMutationId = null,
    mode
  } = input;

  const { cart } = await context.mutations.reconcileCarts(context, {
    anonymousCartId: decodeCartOpaqueId(opaqueAnonymousCartId),
    anonymousCartToken,
    mode
  });

  return { cart, clientMutationId };
}
