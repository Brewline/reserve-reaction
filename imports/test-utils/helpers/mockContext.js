const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
