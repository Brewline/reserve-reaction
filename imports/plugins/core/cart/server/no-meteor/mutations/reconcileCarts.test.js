import mockContext from "/imports/test-utils/helpers/mockContext";
import reconcileCarts from "./reconcileCarts";

jest.mock("./convertAnonymousCartToNewAccountCart", () => jest.fn().mockImplementation(() => Promise.resolve({
  _id: "convertAnonymousCartToNewAccountCart"
})));

jest.mock("./reconcileCartsKeepAccountCart", () => jest.fn().mockImplementation(() => Promise.resolve({
  _id: "reconcileCartsKeepAccountCart"
})));

jest.mock("./reconcileCartsKeepAnonymousCart", () => jest.fn().mockImplementation(() => Promise.resolve({
  _id: "reconcileCartsKeepAnonymousCart"
})));

jest.mock("./reconcileCartsMerge", () => jest.fn().mockImplementation(() => Promise.resolve({
  _id: "reconcileCartsMerge"
})));

const accountId = "accountId";
const anonymousCartId = "anonymousCartId";
const anonymousCartToken = "anonymousCartToken";
const shopId = "shopId";

const mockCarts = [
  { _id: anonymousCartId },
  { _id: "ACCOUNT_CART", accountId }
];

test("when mode is keepAccountCart, returns the result of reconcileCartsKeepAccountCart", async () => {
  mockContext.accountId = accountId;
  mockContext.user = { accountId, roles: ["guest"] };
  mockContext.collections.Cart.toArray.mockReturnValueOnce(Promise.resolve(mockCarts));

  const result = await reconcileCarts(mockContext, {
    anonymousCartId,
    anonymousCartToken,
    mode: "keepAccountCart",
    shopId
  });

  expect(result).toEqual({ cart: { _id: "reconcileCartsKeepAccountCart" } });
});

test("when mode is keepAnonymousCart, returns the result of reconcileCartsKeepAnonymousCart", async () => {
  mockContext.accountId = accountId;
  mockContext.user = { accountId, roles: ["guest"] };
  mockContext.collections.Cart.toArray.mockReturnValueOnce(Promise.resolve(mockCarts));

  const result = await reconcileCarts(mockContext, {
    anonymousCartId,
    anonymousCartToken,
    mode: "keepAnonymousCart",
    shopId
  });

  expect(result).toEqual({ cart: { _id: "reconcileCartsKeepAnonymousCart" } });
});

test("when mode is merge, returns the result of reconcileCartsMerge", async () => {
  mockContext.accountId = accountId;
  mockContext.user = { accountId, roles: ["guest"] };
  mockContext.collections.Cart.toArray.mockReturnValueOnce(Promise.resolve(mockCarts));

  const result = await reconcileCarts(mockContext, {
    anonymousCartId,
    anonymousCartToken,
    mode: "merge",
    shopId
  });

  expect(result).toEqual({ cart: { _id: "reconcileCartsMerge" } });
});

test("when mode is undefined, returns the result of reconcileCartsMerge", async () => {
  mockContext.accountId = accountId;
  mockContext.user = { accountId, roles: ["guest"] };
  mockContext.collections.Cart.toArray.mockReturnValueOnce(Promise.resolve(mockCarts));

  const result = await reconcileCarts(mockContext, {
    anonymousCartId,
    anonymousCartToken,
    shopId
  });

  expect(result).toEqual({ cart: { _id: "reconcileCartsMerge" } });
});

test("when there is no account cart yet, returns the result of convertAnonymousCartToNewAccountCart", async () => {
  mockContext.accountId = accountId;
  mockContext.user = { accountId, roles: ["guest"] };
  mockContext.collections.Cart.toArray.mockReturnValueOnce(Promise.resolve([mockCarts[0]]));

  const result = await reconcileCarts(mockContext, {
    anonymousCartId,
    anonymousCartToken,
    mode: "keepAccountCart",
    shopId
  });

  expect(result).toEqual({ cart: { _id: "convertAnonymousCartToNewAccountCart" } });
});


test("when not authenticated, throw access denied", async () => {
  mockContext.accountId = null;
  mockContext.user = null;

  const promise = reconcileCarts(mockContext, {
    anonymousCartId,
    anonymousCartToken,
    shopId
  });

  return expect(promise).rejects.toThrowErrorMatchingSnapshot();
});

test("when missing anonymousCartId, throws", async () => {
  mockContext.accountId = accountId;
  mockContext.user = { accountId, roles: ["guest"] };

  const promise = reconcileCarts(mockContext, {
    anonymousCartToken,
    shopId
  });

  return expect(promise).rejects.toThrowErrorMatchingSnapshot();
});

test("when missing anonymousCartToken, throws", async () => {
  mockContext.accountId = accountId;
  mockContext.user = { accountId, roles: ["guest"] };

  const promise = reconcileCarts(mockContext, {
    anonymousCartId,
    shopId
  });

  return expect(promise).rejects.toThrowErrorMatchingSnapshot();
});

test("when no matching anonymous cart is found, throws", async () => {
  mockContext.accountId = accountId;
  mockContext.user = { accountId, roles: ["guest"] };
  mockContext.collections.Cart.toArray.mockReturnValueOnce(Promise.resolve([mockCarts[1]]));

  const promise = reconcileCarts(mockContext, {
    anonymousCartId,
    anonymousCartToken,
    shopId
  });

  return expect(promise).rejects.toThrowErrorMatchingSnapshot();
});

test("when unknown mode is passed, throws", async () => {
  mockContext.accountId = accountId;
  mockContext.user = { accountId, roles: ["guest"] };
  mockContext.collections.Cart.toArray.mockReturnValueOnce(Promise.resolve(mockCarts));

  const promise = reconcileCarts(mockContext, {
    anonymousCartId,
    anonymousCartToken,
    mode: "foo",
    shopId
  });

  return expect(promise).rejects.toThrowErrorMatchingSnapshot();
});
