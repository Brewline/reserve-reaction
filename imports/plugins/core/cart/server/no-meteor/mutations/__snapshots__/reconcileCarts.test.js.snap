// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when missing anonymousCartId, throws 1`] = `"anonymousCartId is required"`;

exports[`when missing anonymousCartToken, throws 1`] = `"anonymousCartToken is required"`;

exports[`when no matching anonymous cart is found, throws 1`] = `"Anonymous cart not found"`;

exports[`when not authenticated, throw access denied 1`] = `"Access Denied"`;

exports[`when unknown mode is passed, throws 1`] = `"mode must be keepAccountCart, keepAnonymousCart, or merge"`;
